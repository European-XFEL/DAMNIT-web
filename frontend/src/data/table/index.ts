export {
  DEFERRED_TABLE_DATA_QUERY_NAME,
  LATEST_DATA_FIELD_NAME,
} from './table-data.constants'
export { selectVariables } from './table-data.selectors'
export {
  default as tableDataReducer,
  getTable,
  getTableData,
  resetTable,
  updateTable,
} from './table-data.slice'
export {
  REFRESH_MUTATION,
  LATEST_DATA_SUBSCRIPTION,
} from './table-data.services'
