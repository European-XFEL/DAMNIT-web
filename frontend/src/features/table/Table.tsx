import "@glideapps/glide-data-grid/dist/index.css"

import React, { useCallback, useEffect, useRef, useState } from "react"
import { connect, useDispatch, useSelector } from "react-redux"
import {
  CompactSelection,
  DataEditor,
  GridSelection,
  Item,
} from "@glideapps/glide-data-grid"
import { useExtraCells } from "@glideapps/glide-data-grid-cells"
import { range, useDisclosure } from "@mantine/hooks"

import { gridCellFactory } from "./cells"
import ContextMenu from "./ContextMenu"

import { selectRun } from "./tableSlice"
import { addPlot } from "../plots"
import {
  getExtractedVariable,
  getTableData,
  getTableVariable,
} from "../../redux"

import { DTYPES, VARIABLES } from "../../constants"
import {
  isArrayEqual,
  sorted,
  sortedInsert,
  sortedSearch,
} from "../../utils/array"
import { isDataPlottable } from "../../utils/plots"
import { createMap, isEmpty } from "../../utils/helpers"
import PlotDialog from "../plots/PlotDialog"

export const EXCLUDED_VARIABLES = ["proposal", "start_time"]

class Pages {
  constructor() {
    this.loading = []
    this.loaded = []
  }

  addToLoading(page) {
    sortedInsert(this.loading, page)
  }

  isLoading(page) {
    return sortedSearch(this.loading, page) !== -1
  }

  addToLoaded(page) {
    // Remove from loading
    const index = sortedSearch(this.loading, page)
    if (index !== -1) {
      this.loading.splice(index, 1)
    }

    sortedInsert(this.loaded, page)
  }

  isLoaded(page) {
    return sortedSearch(this.loaded, page) !== -1
  }
}

const usePagination = (proposal, pageSize = 5) => {
  const dispatch = useDispatch()

  // Reference: Loaded pages
  const pagesRef = useRef(new Pages())

  // State: Visible pages
  const [visibleRegion, setVisibleRegion] = useState({
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  })

  // Callback: Handle new page
  const handleNewPage = useCallback(
    async (page, pageSize) => {
      if (!proposal) {
        return
      }

      let loaded = false

      if (page > 0) {
        // TODO: Create an object that contains `page` and `pageSize`
        loaded = dispatch(getTableData({ proposal, page })).then(() => true)
      }

      return loaded
    },
    [proposal, dispatch],
  )

  // Callback: On visible region changed
  const onVisibleRegionChanged = useCallback((rect) => {
    setVisibleRegion((cv) => {
      return rect
    })
  }, [])

  const loadPage = useCallback(
    async (page: number) => {
      // TODO: Add a retry when loading pages are stuck for quite some time
      const pages = pagesRef.current

      if (pages.isLoading(page) || pages.isLoaded(page)) {
        return
      }

      pages.addToLoading(page)
      const loaded = await handleNewPage(page, pageSize)
      if (loaded) {
        pages.addToLoaded(page)
      }
    },
    [handleNewPage, pageSize],
  )

  // Effect: Trigger load page when visible region changes
  useEffect(() => {
    if (visibleRegion.width === 0 || visibleRegion.height === 0) {
      return
    }

    const firstPage = Math.max(
      0,
      Math.floor((visibleRegion.y - pageSize / 2) / pageSize),
    )
    const lastPage = Math.floor(
      (visibleRegion.y + visibleRegion.height + pageSize / 2) / pageSize,
    )
    range(firstPage + 1, lastPage + 2).map((page) => {
      if (!pagesRef.current.isLoaded(page)) {
        loadPage(page)
      }
    })
  }, [loadPage, pageSize, visibleRegion])

  return { onVisibleRegionChanged }
}

const useContextMenu = () => {
  const closedProps = {
    localPosition: { x: 0, y: 0 },
    bounds: {},
    isOpen: false,
    contents: [],
  }
  const [props, setProps] = useState(closedProps)

  const handleOutsideClick = () => {
    setProps(closedProps)
  }

  const updateProps = (newProps) => {
    const updated = newProps ? { ...newProps, isOpen: true } : closedProps
    setProps(updated)
  }

  return [{ ...props, onOutsideClick: handleOutsideClick }, updateProps]
}

const Table = (props) => {
  // Initialization
  const cellProps = useExtraCells()
  const proposal = useSelector((state) => state.proposal.current.value)
  const paginationProps = usePagination(proposal)

  // Data: Populate grid
  const getContent = useCallback(
    ([col, row]) => {
      const column = props.columns[col].id
      const rowData = props.data[row]

      if (!rowData || !rowData[column]) {
        return gridCellFactory[DTYPES.string]("")
      }

      const cell = gridCellFactory[rowData[column]?.dtype]
      return cell(rowData[column].value, {
        lastUpdated: props.lastUpdate[rowData[VARIABLES.run_number]],
      })
    },
    [props.columns, props.data, props.lastUpdate],
  )

  // Cell: Click event
  const [gridSelection, setGridSelection] = useState({
    columns: CompactSelection.empty(),
    rows: CompactSelection.empty(),
  })
  const handleGridSelectionChange = (newSelection: GridSelection) => {
    const { columns, rows, current } = newSelection

    // Inform that a row has been (de)selected
    const row = rows.last()
    props.dispatch(
      selectRun({
        run:
          isEmpty(row) || !props.data[row]
            ? null
            : props.data[row][VARIABLES.run_number].value,
      }),
    )

    // Clear range stack if cells from the other column are currently selected
    let rangeStack
    if (!isEmpty(current?.cell) && !isEmpty(current?.rangeStack)) {
      rangeStack =
        current.cell[0] != current.rangeStack[0].x
          ? []
          : current?.rangeStack.filter(
              (range) => !isArrayEqual(current.cell, [range.x, range.y]),
            )
    }

    // Finalize
    setGridSelection({
      columns,
      rows,
      ...(current && {
        current: { ...current, ...(rangeStack && { rangeStack }) },
      }),
    })
  }
  const handleCellActivated = (cell: Item) => {
    const [col, row] = cell
    props.dispatch(
      selectRun({
        run: isEmpty(row) ? null : props.data[row][VARIABLES.run_number].value,
        variables: isEmpty(col) ? null : [props.columns[col].id],
      }),
    )
  }

  // Context menus
  const [contextMenu, setContextMenu] = useContextMenu()
  const handleCellContextMenu = ([col, row], event) => {
    event.preventDefault()

    const column = props.columns[col].id
    const rowData = props.data[row]

    if (col !== -1 && isDataPlottable(rowData[column].dtype)) {
      setContextMenu({
        localPosition: { x: event.localEventX, y: event.localEventY },
        bounds: event.bounds,
        contents: ["plot"],
      })
    }
  }
  const handleHeaderContextMenu = (col, event) => {
    event.preventDefault()
    const columnSelection = gridSelection.columns.toArray()
    if (columnSelection.length === 1) {
      if (col !== -1) {
        setContextMenu({
          localPosition: { x: event.localEventX, y: event.localEventY },
          bounds: event.bounds,
          contents: ["plot"],
        })
      }
    } else if (columnSelection.length === 2) {
      setContextMenu({
        localPosition: { x: event.localEventX, y: event.localEventY },
        bounds: event.bounds,
        contents: ["corrPlot"],
      })
    }
  }

  const handleAddPlot = () => {
    if (gridSelection.current) {
      const { cell, rangeStack } = gridSelection.current
      const rows = [cell[1], ...rangeStack.map((stack) => stack.y)]

      const variable = props.columns[cell[0]].id
      const runs = sorted(
        rows.map((row) => props.data[row][VARIABLES.run_number].value),
      )
      props.dispatch(
        addPlot({
          runs,
          variables: [variable],
          source: "extracted",
        }),
      )
      runs.forEach((run) => {
        props.dispatch(getExtractedVariable({ proposal, run, variable }))
      })
    } else {
      const col = gridSelection.columns.last()
      props.dispatch(
        addPlot({
          variables: [props.columns[col].id],
          source: "table",
        }),
      )
      props.dispatch(
        getTableVariable({
          proposal,
          variable: props.columns[col].id,
        }),
      )
    }
  }

  const handleAddCorrPlot = () => {
    const cols = gridSelection.columns.toArray()
    props.dispatch(
      addPlot({
        variables: [props.columns[cols[0]].id, props.columns[cols[1]].id],
        source: "table",
      }),
    )
  }

  const handlePlotOptions = () => {
    // set states for pre configurable plotting settings here
    // Pass they as props to the modal
    handlersPlotDialog.open()
  }

  const cellContextContents = createMap(
    [
      {
        key: "plot",
        title: "Plot",
        onClick: handleAddPlot,
      },
      {
        key: "corrPlot",
        title: "Correlation plot",
        onClick: handleAddCorrPlot,
      },
      {
        key: "advPlot",
        title: "Plot options",
        onClick: handlePlotOptions,
      },
    ],
    "key",
  )

  const [openedPlotDialog, handlersPlotDialog] = useDisclosure(false)

  return (
    <div style={{ width: "100%", height: "100%" }}>
      {!props.columns.length ? null : (
        <>
          <DataEditor
            {...(props.grid || {})}
            columns={formatColumns(props.columns)}
            getCellContent={getContent}
            rows={props.rows}
            rowSelect="single"
            rowMarkers="clickable-number"
            gridSelection={gridSelection}
            onGridSelectionChange={handleGridSelectionChange}
            onCellActivated={handleCellActivated}
            rangeSelect="multi-cell"
            onCellContextMenu={handleCellContextMenu}
            onHeaderContextMenu={handleHeaderContextMenu}
            freezeColumns={1}
            {...cellProps}
            {...paginationProps}
          />
          <ContextMenu
            {...contextMenu}
            contents={contextMenu.contents.map((key) =>
              cellContextContents.get(key),
            )}
          />
          {openedPlotDialog && (
            <PlotDialog
              opened={openedPlotDialog}
              open={handlersPlotDialog.open}
              close={handlersPlotDialog.close}
              selectedColumns={props.columns
                .filter((_, index) =>
                  gridSelection.columns.toArray().includes(index),
                )
                .map((col) => col.id)}
            />
          )}
          <div id="portal" />
        </>
      )}
    </div>
  )
}

const mapStateToProps = ({ tableData: table }) => {
  const data = table.data ? Object.values(table.data) : []
  // TODO: Get the column list from the user settings (reordered columns)
  const columns = Object.values(table.metadata.variables)
    .filter(({ name }) => !EXCLUDED_VARIABLES.includes(name))
    .map(({ name, title }) => ({ id: name, title }))

  return {
    data,
    columns,
    rows: table.metadata.rows,
    lastUpdate: table.lastUpdate,
  }
}

export default connect(mapStateToProps)(Table)

const formatColumns = (columns) => {
  return columns.map((column) => ({
    ...column,
    width: 100,
  }))
}
