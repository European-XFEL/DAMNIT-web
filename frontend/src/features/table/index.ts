export { default } from "./Table"
export {
  default as tableReducer,
  selectRun,
  reset as resetTable,
} from "./tableSlice"
