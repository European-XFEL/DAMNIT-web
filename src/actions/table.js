import { tableConstants } from "../constants";

export const tableActions = {
  receiveTable,
  selectRow,
};

function receiveTable(table) {
  return {
    type: tableConstants.RECEIVE_TABLE,
    payload: table,
  };
}

function selectRow(row) {
  return {
    type: tableConstants.SELECT_ROW,
    payload: { selection: { row } },
  };
}
