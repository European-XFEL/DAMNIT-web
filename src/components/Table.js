import "@glideapps/glide-data-grid/dist/index.css";

import React, { useCallback } from "react";
import { connect } from "react-redux";
import { DataEditor, GridCellKind } from "@glideapps/glide-data-grid";
import { useExtraCells } from "@glideapps/glide-data-grid-cells";

import { imageBytesToURL } from "../utils/helpers";

const EMPTY_VALUE = "None";

const imageCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Image,
    allowOverlay: true,
    data: [imageBytesToURL(data)],
  };
};

const textCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Text,
    allowOverlay: false,
    displayData: data !== EMPTY_VALUE ? data : "",
    data,
  };
};

const numberCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Number,
    allowOverlay: false,
    displayData: data !== EMPTY_VALUE ? String(data) : "",
    data,
  };
};

const arrayCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Custom,
    allowOverlay: false,
    copyData: "4",
    data: {
      kind: "sparkline-cell",
      values: data,
      // displayValues: data.map((x) => Math.round(x).toString()), // TODO: Round in server?
      color: "#77c4c4",
      yAxis: [Math.min(...data), Math.max(...data)],
    },
  };
};

const gridCellFactory = {
  image: imageCell,
  string: textCell,
  number: numberCell,
  array: arrayCell,
};

const Table = ({ data, columns, schema }) => {
  const cellProps = useExtraCells();

  const getContent = useCallback(([col, row]) => {
    const column = columns[col].id;
    const rowData = data[row];
    return gridCellFactory[schema[column]](rowData[column]);
  }, []);

  return (
    <div>
      <DataEditor
        columns={columns}
        getCellContent={getContent}
        rows={data.length}
        {...cellProps}
      />
      <div id="portal"></div>
    </div>
  );
};

const mapStateToProps = ({ table }) => {
  const data = table.data ? Object.values(table.data) : [];

  // TODO: Get the column list from the user settings (reordered columns)
  const columns = table.schema
    ? Object.keys(table.schema).map((id) => ({ id, title: id }))
    : [];

  return { data, columns, schema: table.schema };
};

export default connect(mapStateToProps)(Table);
