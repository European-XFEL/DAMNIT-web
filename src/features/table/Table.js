import "@glideapps/glide-data-grid/dist/index.css";

import React, { useCallback } from "react";
import { connect } from "react-redux";
import { DataEditor, GridCellKind } from "@glideapps/glide-data-grid";
import { useExtraCells } from "@glideapps/glide-data-grid-cells";

import { selectRow } from "./tableSlice";
import { imageBytesToURL } from "../../utils/helpers";

const EMPTY_VALUE = "None";

const imageCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Image,
    allowOverlay: true,
    data: [imageBytesToURL(data)],
  };
};

const textCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Text,
    allowOverlay: false,
    displayData: data !== EMPTY_VALUE ? data : "",
    data,
  };
};

const numberCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Number,
    allowOverlay: false,
    displayData: data !== EMPTY_VALUE ? String(data) : "",
    data,
  };
};

const arrayCell = (data, params = {}) => {
  return {
    kind: GridCellKind.Custom,
    allowOverlay: false,
    copyData: "4",
    data: {
      kind: "sparkline-cell",
      values: data,
      // displayValues: TODO: Round in server?
      color: "#77c4c4",
      yAxis: [Math.min(...data), Math.max(...data)],
    },
  };
};

const gridCellFactory = {
  image: imageCell,
  string: textCell,
  number: numberCell,
  array: arrayCell,
};

const Table = ({ data, columns, schema, selection, dispatch }) => {
  // Initialization: Use custom cells
  const cellProps = useExtraCells();

  // Data: Populate grid
  const getContent = useCallback(
    ([col, row]) => {
      const column = columns[col].id;
      const rowData = data[row];
      return gridCellFactory[schema[column].dtype](rowData[column]);
    },
    [columns, data]
  );

  // Cell: Click event
  const handleCellClicked = ([col, row], params) => {
    // Only handle handles click on the row markers (col = -1)
    if (col !== -1) {
      return;
    }

    // Inform that the row has been (de)selected
    dispatch(selectRow(row === selection.row ? null : row));
  };

  return (
    <div>
      {columns.length && (
        <>
          <DataEditor
            columns={columns}
            getCellContent={getContent}
            rows={data.length}
            rowSelect="single"
            rowMarkers="clickable-number"
            onCellClicked={handleCellClicked}
            {...cellProps}
          />
          <div id="portal"></div>
        </>
      )}
    </div>
  );
};

const mapStateToProps = ({ table }) => {
  const data = table.data ? Object.values(table.data) : [];

  // TODO: Get the column list from the user settings (reordered columns)
  const columns = Object.keys(table.schema).map((id) => ({ id, title: id }));

  return {
    data,
    columns,
    schema: table.schema,
    selection: table.selection ? table.selection : {},
  };
};

export default connect(mapStateToProps)(Table);
