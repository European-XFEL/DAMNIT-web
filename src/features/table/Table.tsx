import "@glideapps/glide-data-grid/dist/index.css"

import React, { useCallback, useState } from "react"
import { connect } from "react-redux"
import {
  CompactSelection,
  DataEditor,
  GridSelection,
} from "@glideapps/glide-data-grid"
import { useExtraCells } from "@glideapps/glide-data-grid-cells"

import { gridCellFactory } from "./cells"
import ContextMenu from "./ContextMenu"

import { selectRun } from "./tableSlice"
import { addPlot } from "../plots"

import { DTYPES, VARIABLES } from "../../common/constants"
import { arrayEqual, isEmpty } from "../../utils/helpers"

const Table = (props) => {
  // Initialization: Use custom cells
  const cellProps = useExtraCells()

  // Data: Populate grid
  const getContent = useCallback(
    ([col, row]) => {
      const column = props.columns[col].id
      const rowData = props.data[row]
      return gridCellFactory[props.schema[column].dtype](rowData[column])
    },
    [props.columns, props.data],
  )

  // Cell: Click event
  const [gridSelection, setGridSelection] = useState({
    columns: CompactSelection.empty(),
    rows: CompactSelection.empty(),
  })
  const onGridSelectionChange = (newSelection: GridSelection) => {
    const { columns, rows, current } = newSelection

    // Inform that a row has been (de)selected
    const row = rows.last()
    props.dispatch(
      selectRun(isEmpty(row) ? null : props.data[row][VARIABLES.run_number]),
    )

    // Clear range stack if cells from the other column are currently selected
    let rangeStack
    if (!isEmpty(current?.cell) && !isEmpty(current?.rangeStack)) {
      rangeStack =
        current.cell[0] != current.rangeStack[0].x
          ? []
          : current?.rangeStack.filter(
              (range) => !arrayEqual(current.cell, [range.x, range.y]),
            )
    }

    // Finalize
    setGridSelection({
      columns,
      rows,
      ...(current && {
        current: { ...current, ...(rangeStack && { rangeStack }) },
      }),
    })
  }

  // Context menus
  const [contextMenu, setContextMenu] = useState()
  const handleCellContextMenu = ([col, row], event) => {
    event.preventDefault()

    // Ignore no selection for the meantime
    if (!gridSelection.current) {
      return
    }

    const cells = [
      gridSelection.current.cell,
      ...gridSelection.current.rangeStack.map((range) => [range.x, range.y]),
    ]

    if (
      col !== -1 &&
      cells.length > 1 &&
      cells.some((cell) => arrayEqual(cell, [col, row])) &&
      props.schema[props.columns[col].id].dtype === DTYPES.number
    ) {
      setContextMenu({
        localPosition: { x: event.localEventX, y: event.localEventY },
        bounds: event.bounds,
      })
    }
  }
  const handleHeaderContextMenu = (col, event) => {
    event.preventDefault()
    if (col !== -1) {
      setGridSelection({
        columns: CompactSelection.fromSingleSelection(col),
        rows: CompactSelection.empty(),
      })

      if (props.schema[props.columns[col].id].dtype === DTYPES.number) {
        setContextMenu({
          localPosition: { x: event.localEventX, y: event.localEventY },
          bounds: event.bounds,
        })
      }
    }
  }
  const handleAddPlot = () => {
    if (gridSelection.current) {
      const { cell, rangeStack } = gridSelection.current
      const rows = [cell[1], ...rangeStack.map((stack) => stack.y)]
      props.dispatch(
        addPlot({
          variables: [props.columns[cell[0]].id],
          runs: rows
            .map((row) => props.data[row][VARIABLES.run_number])
            .toSorted(),
        }),
      )
    } else {
      const col = gridSelection.columns.last()
      props.dispatch(
        addPlot({
          variables: [props.columns[col].id],
          runs: props.data
            .map((rowData) => rowData[VARIABLES.run_number])
            .toSorted(),
        }),
      )
    }
  }
  const cellContextContents = [
    {
      key: "plot",
      title: "Plot",
      onClick: handleAddPlot,
    },
  ]

  // Format columns
  const formattedColumns = props.columns.map((column) => ({
    ...column,
    ...(props.schema[column.id].dtype === DTYPES.number && {
      width: 100,
      themeOverride: {
        fontFamily: "monospace",
        headerFontStyle: "",
      },
    }),
  }))

  return (
    <div>
      {!props.columns.length ? null : (
        <>
          <DataEditor
            {...(props.grid || {})}
            id="main-table"
            columns={formattedColumns}
            getCellContent={getContent}
            rows={props.data.length}
            rowSelect="single"
            rowMarkers="clickable-number"
            gridSelection={gridSelection}
            onGridSelectionChange={onGridSelectionChange}
            rangeSelect="multi-cell"
            onCellContextMenu={handleCellContextMenu}
            onHeaderContextMenu={handleHeaderContextMenu}
            freezeColumns={1}
            {...cellProps}
          />
          {contextMenu && (
            <ContextMenu
              {...contextMenu}
              onOutsideClick={() => setContextMenu(null)}
              contents={cellContextContents}
            />
          )}
          <div id="portal"></div>
        </>
      )}
    </div>
  )
}

const mapStateToProps = ({ table }) => {
  const data = table.data ? Object.values(table.data) : []

  // TODO: Get the column list from the user settings (reordered columns)
  const columns = Object.keys(table.schema)
    .filter((id) => id !== VARIABLES.proposal)
    .map((id) => ({ id, title: id }))

  return {
    data,
    columns,
    schema: table.schema,
  }
}

export default connect(mapStateToProps)(Table)
