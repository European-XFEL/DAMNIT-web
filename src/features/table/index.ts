export { default } from "./Table";
export { default as tableReducer, getTable, selectRun } from "./tableSlice";
