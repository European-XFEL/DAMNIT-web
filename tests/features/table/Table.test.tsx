import React from "react"
import Table from "@/features/table/Table"

import { renderWithProviders, screen } from "../../test-utils/extensions"
import { gridProps, validTable } from "../../test-utils/builders/table"

beforeEach(() => {
  Element.prototype.scrollTo = vi.fn()
  Element.prototype.scrollBy = vi.fn()
  Element.prototype.getBoundingClientRect = () => ({
    bottom: 1000,
    height: 1000,
    left: 0,
    right: 1000,
    top: 0,
    width: 1000,
    x: 0,
    y: 0,
    toJSON: () => "",
  })
  Image.prototype.decode = vi.fn()
})

afterEach(() => {
  // vi.clearAllTimers()
  vi.clearAllMocks()
})

describe("Table", () => {
  it("does not render canvas with invalid input", () => {
    renderWithProviders(<Table grid={gridProps} />)
    expect(screen.queryByTestId("data-grid-canvas")).not.toBeInTheDocument()
  })

  it("renders canvas with valid input", async () => {
    renderWithProviders(<Table grid={gridProps} />, {
      preloadedState: {
        table: validTable,
      },
    })

    expect(screen.getByTestId("data-grid-canvas")).toBeInTheDocument()
  })
})
