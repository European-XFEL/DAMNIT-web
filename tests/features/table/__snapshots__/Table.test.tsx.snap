// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should draw a house on the canvas using the main ctx 1`] = `
[
  {
    "props": {
      "value": "#808080",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  {
    "props": {
      "fillRule": "nonzero",
      "path": [],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  {
    "props": {
      "fillRule": "nonzero",
      "path": [],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  {
    "props": {
      "value": "#ffffff",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  {
    "props": {
      "fillRule": "nonzero",
      "path": [],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  {
    "props": {
      "value": 2,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  {
    "props": {
      "value": "#ffff00",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  {
    "props": {
      "path": [],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  {
    "props": {
      "path": [],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
